Code quality refactor of PlayerService and FilteringService

hburn7 opened the pull request:
Focused on quality, not efficiency. Filtering functionality still works great.

Closes #305

myssto commented on API/Services/Implementations/FilteringService.cs:
```diff
@@ -12,113 +12,118 @@ public class FilteringService(
+        };
+    }
+
+    private async Task<PlayerFilteringResultDTO> ProcessPlayerAsync(
+        PlayerCompactDTO? playerInfo,
+        FilteringRequestDTO request)
+    {
+        (FilteringResult result, FilteringFailReason? failReason) = await FilterAsync(request, playerInfo);
+
+        return new PlayerFilteringResultDTO
+        {
+            PlayerId = playerInfo?.Id,
+            Username = playerInfo?.Username,
+            OsuId = playerInfo?.OsuId ?? 0,
+            FilteringResult = result,
+            FilteringFailReason = failReason
         };
     }
```
don't understand why this was moved into it's own function when it is only used once. would rather this just be inlined where it was before

myssto commented on API/Services/Implementations/FilteringService.cs:
```diff
@@ -12,113 +12,118 @@ public class FilteringService(
+            results.Add(result);
         }
 
+        // Count results and create result DTO
+        var passed = results.Count(r => r.FilteringResult == FilteringResult.Pass);
+        var failed = results.Count(r => r.FilteringResult == FilteringResult.Fail);
+
         return new FilteringResultDTO
         {
             PlayersPassed = passed,
             PlayersFailed = failed,
```
```suggestion
        return new FilteringResultDTO
        {
            PlayersPassed = results.Count(r => r.FilteringResult == FilteringResult.Pass),
            PlayersFailed = results.Count(r => r.FilteringResult == FilteringResult.Fail),
```

myssto commented on API/Services/Implementations/FilteringService.cs:
```diff
@@ -12,113 +12,118 @@ public class FilteringService(
         if (idList.Count == 0)
         {
-            throw new Exception("Filtering id list cannot be empty");
+            throw new ArgumentException("Filtering id list cannot be empty", nameof(filteringRequest));
```
don't see the need for this to throw considering none of the operations depending on this list will cause an unrecoverable error if it is empty. this should probably be validated at the controller, or better yet with a data annotation on the request body which cause it the request to fail at validation time

myssto commented on API/Services/Implementations/FilteringService.cs:
```diff
@@ -12,113 +12,118 @@ public class FilteringService(
 {
     public async Task<FilteringResultDTO> FilterAsync(FilteringRequestDTO filteringRequest)
     {
+        ArgumentNullException.ThrowIfNull(filteringRequest);
```
when will this ever have a chance of being null? I understand this is common practice in c# but we never do this anywhere in the codebase

hburn7 replied:
Eh yeah I shouldn't have put this here since it's not a nullable type, will remove.

myssto commented on API/Services/Implementations/FilteringService.cs:
```diff
@@ -12,113 +12,118 @@ public class FilteringService(
         var idList = filteringRequest.OsuPlayerIds.ToList();
 
         if (idList.Count == 0)
         {
-            throw new Exception("Filtering id list cannot be empty");
+            throw new ArgumentException("Filtering id list cannot be empty", nameof(filteringRequest));
         }
 
-        var passed = 0;
-        var failed = 0;
+        IEnumerable<PlayerCompactDTO?> players = await playerService.GetAsync(idList);
```
don't really need to define `idList` as it's own variable, just pass it to this function like `playerService.GetAsync(filteringRequest.OsuPlayerIds)`

myssto commented on API/Services/Implementations/FilteringService.cs:
```diff
@@ -12,113 +12,118 @@ public class FilteringService(
         }
 
-        if (ratingStats.Rating < filteringRequest.MinRating)
+        FilteringResult result = failReason == FilteringFailReason.None ? FilteringResult.Pass : FilteringResult.Fail;
```
wait so why do even have an enum `FilteringResult` for this when it literally represents a boolean `pass` or `fail` LOL? This function should just return a `FilteringFailReason` and whether or not filtering failed should be implicit based on the flags

myssto commented on API/Services/Implementations/FilteringService.cs:
```diff
@@ -12,113 +12,118 @@ public class FilteringService(
         if (playerInfo == null)
         {
-            failReason |= FilteringFailReason.NoData;
-
-            return (result, failReason);
+            return (FilteringResult.Fail, FilteringFailReason.NoData);
         }
 
-        PlayerRatingStatsDTO? ratingStats =
-            await playerRatingsService.GetAsync(playerInfo.Id, filteringRequest.Ruleset, includeAdjustments: false);
+        PlayerRatingStatsDTO? ratingStats = await playerRatingsService.GetAsync(
+            playerInfo.Id,
+            request.Ruleset,
+            includeAdjustments: false);
 
         if (ratingStats == null)
         {
-            failReason |= FilteringFailReason.NoData;
+            return (FilteringResult.Fail, FilteringFailReason.NoData);
+        }
 
-            return (result, failReason);
+        FilteringFailReason failReason = CheckFilteringConditions(request, ratingStats);
+
+        var peakRating = await playerStatsService.GetPeakRatingAsync(playerInfo.Id, request.Ruleset);
+        if (peakRating > request.PeakRating)
+        {
+            failReason |= FilteringFailReason.PeakRatingTooHigh;
         }
 
-        if (ratingStats.Rating < filteringRequest.MinRating)
+        FilteringResult result = failReason == FilteringFailReason.None ? FilteringResult.Pass : FilteringResult.Fail;
+
+        return (result, failReason);
+    }
```
Why are this function and `CheckFilteringConditions` separate? Why does `CheckFilteringConditions` only return a `FilteringFailReason` and not a tuple like this one does (which it shouldn't in the first place)?

myssto reviewed (changes requested):
Will review rest after changes

hburn7 commented:
- Refactored a lot of the AI jank
- Created a new `SupportedRuleset` validation attribute to handle the case where `osu!mania (Other)` is provided. I think we could use this as a data annotation on all of our DTOs which have a Ruleset field.
- Replaced the `FilteringResult` enum with a boolean that switches based on the state of the failure reason.

myssto commented on API/Services/Implementations/PlayerService.cs:
```diff
@@ -8,45 +8,24 @@ namespace API.Services.Implementations;
+            .Where(p => p is not null)
+            .ToDictionary(p => p!.OsuId);  // Assuming Player has OsuId property
+
+        // Return a list matching the input order:
+        // - If the player exists in our dictionary, return the mapped DTO
+        // - If not found, return null
+        return idList.Select(id =>
+            players.TryGetValue(id, out Player? player)
+                ? mapper.Map<PlayerCompactDTO>(player)
+                : null);
```
Is there a reason that this very specific return format is done here instead of where the data is actually being used? seems a bit strange considering this is probably only used in one place right?

myssto replied:
to be more specific, I don't necessarily see why it should be the responsibility of this service to provide a null in place when a player is not found out of a list of ids. its not consistent with any other impl of `IEnumerable<Entity> GetAsync(IEnumerable<int> ids)` in the codebase. caller of the function should be able to assume that if no player was returned for an id that it input, one does not exist.

hburn7 replied:
I was mainly trying to avoid looping through the list a second time to determine whether anyone is missing. In practice that is kinda jank and that's what a Dictionary is for. Even with a Dictionary impl, another iteration would need to be done somewhere to identify users who weren't fetched.

I think it's important to not leave room for assumptions in this case (from the perspective of the API caller not knowing why there are missing results). We can include a list of IDs which returned no result instead of including nulls in the output. I'll redo this impl with that in mind.

myssto commented on API/DTOs/FilteringRequestDTO.cs:
```diff
@@ -59,5 +60,6 @@ public class FilteringRequestDTO
     /// A list of osu! player ids that will be filtered
     /// </summary>
-    public required IEnumerable<long> OsuPlayerIds { get; set; } = [];
+    [MinLength(1, ErrorMessage = "At least one osu! player ID is required.")]
```
```suggestion
    [MinLength(1)]
```
I have tried to avoid using custom error messages if at all possible

myssto commented on API/DTOs/PlayerFilteringResultDTO.cs:
```diff
@@ -1,38 +1,51 @@
-    public string FilteringResultMessage => FilteringResult.ToString();
+
     /// <summary>
     /// The <see cref="FilteringFailReason"/> in string form
     /// </summary>
-    public string? FilteringFailReasonMessage => FilteringFailReason.ToString();
+    public string? FilteringFailReasonMessage => FilteringFailReason is null ? null : FilteringFailReasonMessage;
```
do we even really need this? we use raw enums in every other DTO, don't see how this one should be treated different

myssto commented on API/DTOs/PlayerFilteringResultDTO.cs:
```diff
@@ -1,38 +1,51 @@
+
     /// <summary>
-    /// The filtering result
+    /// Whether the player successfully passed through the filter.
+    /// If this value is true, the player meets all constraints
+    /// specified in the filter.
```
```suggestion
    /// Whether the player successfully passes all conditions of the filter
```

myssto commented on API/DTOs/PlayerFilteringResultDTO.cs:
```diff
@@ -1,38 +1,51 @@
+    public bool IsSuccess { get; set; }
+
     /// <summary>
     /// If the user failed filtering, the fail reason
```
```suggestion
    /// Reason for failure if applicable
```

myssto commented on API/Services/Implementations/FilteringService.cs:
```diff
@@ -10,115 +10,104 @@ public class FilteringService(
-                default:
-                    throw new Exception("This FilteringResult is not handled!");
-            }
-
-            resultCollection.Add(new PlayerFilteringResultDTO
-            {
-                PlayerId = playerInfo.Id,
-                Username = playerInfo.Username,
-                OsuId = playerInfo.OsuId,
-                FilteringResult = result,
+                PlayerId = playerInfo?.Id,
+                Username = playerInfo?.Username,
+                OsuId = playerInfo?.OsuId ?? 0,
                 FilteringFailReason = failReason
             });
         }
```
this could probably be done nicely with a select statement

hburn7 replied:
I originally tried with a select statement but it ends up returning a task list. Even with an async lambda, I can't await each filter result immediately.

myssto replied:
that's true, didn't think of that and they aren't safe to multithread because of db interface

myssto commented on API/Services/Implementations/FilteringService.cs:
```diff
@@ -10,115 +10,104 @@ public class FilteringService(
-        if (ratingStats.Rating > filteringRequest.MaxRating)
+        if (ratingStats.Rating > request.MaxRating)
         {
             failReason |= FilteringFailReason.MaxRating;
         }
 
-        if (!filteringRequest.AllowProvisional && ratingStats.IsProvisional)
+        if (!request.AllowProvisional && ratingStats.IsProvisional)
         {
             failReason |= FilteringFailReason.IsProvisional;
         }
 
-        if (ratingStats.MatchesPlayed < filteringRequest.MatchesPlayed)
-        {
-            failReason |= FilteringFailReason.NotEnoughMatches;
-        }
-
-        if (ratingStats.TournamentsPlayed < filteringRequest.TournamentsPlayed)
+        if (ratingStats.TournamentsPlayed < request.TournamentsPlayed)
         {
             failReason |= FilteringFailReason.NotEnoughTournaments;
         }
 
-        var peakRating = await playerStatsService.GetPeakRatingAsync(playerInfo.Id, filteringRequest.Ruleset);
-        if (peakRating > filteringRequest.PeakRating)
+        if (peakRating > request.PeakRating)
         {
             failReason |= FilteringFailReason.PeakRatingTooHigh;
         }
 
-        if (failReason == FilteringFailReason.None)
+        if (ratingStats.MatchesPlayed < request.MatchesPlayed)
         {
-            result = FilteringResult.Pass;
+            failReason |= FilteringFailReason.NotEnoughMatches;
         }
 
-        return (result, failReason);
+        return failReason;
     }
```
these two functions should just be merged

hburn7 replied:
I disagree, I think these should be split because this is the single location where all filtering enforcement will live (separation of concerns). This is split on the same grounds as `FilterPlayerAsync`'s logic being separated from `FilterAsync`.

myssto replied:
dont really have a strong opinion either way

myssto commented on Common/Enums/FilteringFailReason.cs:
```diff
@@ -1,43 +1,51 @@
 {
     /// <summary>
-    /// The player passed filtering, thus there is no failure reason
+    /// The player does not exist in the system at all
```
```suggestion
    /// The player is not indexed by o!TR
```
reads a bit nicer maybe?

myssto commented:
> I was mainly trying to avoid looping through the list a second time to determine whether anyone is missing. In practice that is kinda jank and that's what a Dictionary is for. Even with a Dictionary impl, another iteration would need to be done somewhere to identify users who weren't fetched.

I understand and respect wanting to preemptively optimize but to be frank I seriously think it just doesn't matter. I would rather things be a bit easier to follow and work with than creating incredibly specific single-use methods on a generic repository. Imo doing it the way it is currently with the `NotFoundOsuIds` list in the response is silly and doesn't align with anything else in the codebase. And fwiw, `NotFoundOsuIds = [.. osuIdHashSet.Except(players.Select(p => p.OsuId))]` is going to cause that extra enumeration to occur that you were trying to avoid in the first place. At that point why not just take the resulting difference and map them to a result anyway?

 ```cs
osuIdHashSet
  .Except(players.Select(p => p.OsuId))
  .Select(osuId => {
      return new PlayerFilteringResultDTO
      {
          OsuId = osuId,
          IsSuccess = false,
          FailureReason = FilteringFailReason.NoData
      }
  })
```

I think you should just:
- Receive `IEnumberable<long> osuIds` from filter request
- Get `IEnumerable<Player> players` from database
- Iterate over `osuIds`, **not** `players`
- Determine if player is missing with `players.FirstOrDefault(p => p.OsuId == osuId)`

tada every osu id in the request gets a result
